src/utilities/findBreakingChanges.js

```diff
@@ -278,22 +354,128 @@ export function findFieldsThatChangedType(
+      ) {
+        breakingFieldChanges.push({
+          type: BreakingChangeType.NON_NULL_INPUT_FIELD_ADDED,
+          description: `A non-null field ${fieldName} on ` +
+            `input type ${newType.name} was added.`,
+        });
+      }
+    });
   });
   return breakingFieldChanges;
 }
 
+function isChangeSafeForObjectOrInterfaceField(
+  oldType: GraphQLType,
+  newType: GraphQLType,
+): boolean {
+  let isSafe = false;
+  if (isNamedType(oldType)) {
+    isSafe =
+      // if they're both named types, see if their names are equivalent
+      (
+        isNamedType(newType) && oldType.name === newType.name
+      ) ||
+      // moving from nullable to non-null of the same underlying type is safe
+      (
+        newType instanceof GraphQLNonNull &&
+        isChangeSafeForObjectOrInterfaceField(
+          oldType,
+          newType.ofType,
+        )
+      );
+  } else if (oldType instanceof GraphQLList) {
+    isSafe =
+      // if they're both lists, make sure the underlying types are compatible
+      (
+        newType instanceof GraphQLList &&
+        isChangeSafeForObjectOrInterfaceField(
+          oldType.ofType,
+          newType.ofType,
+        )
+      ) ||
+      // moving from nullable to non-null of the same underlying type is safe
+      (
+        newType instanceof GraphQLNonNull &&
+        isChangeSafeForObjectOrInterfaceField(
+          oldType,
+          newType.ofType,
+        )
+      );
+  } else if (oldType instanceof GraphQLNonNull) {
+    // if they're both non-null, make sure the underlying types are compatible
+    isSafe = newType instanceof GraphQLNonNull &&
+      isChangeSafeForObjectOrInterfaceField(
+        oldType.ofType,
+        newType.ofType,
+      );
+  }
+  return isSafe;
+}
+
+function isChangeSafeForInputObjectFieldOrFieldArg(
+  oldType: GraphQLType,
+  newType: GraphQLType,
+): boolean {
+  let isSafe = false;
+  if (isNamedType(oldType)) {
+    // if they're both named types, see if their names are equivalent
+    isSafe = isNamedType(newType) && oldType.name === newType.name;
+  } else if (oldType instanceof GraphQLList) {
+    // if they're both lists, make sure the underlying types are compatible
+    isSafe = newType instanceof GraphQLList &&
+      isChangeSafeForInputObjectFieldOrFieldArg(
+        oldType.ofType,
+        newType.ofType,
+      );
+  } else if (oldType instanceof GraphQLNonNull) {
+    isSafe =
+    // if they're both non-null, make sure the underlying types are compatible
+      (
+        newType instanceof GraphQLNonNull &&
+        isChangeSafeForInputObjectFieldOrFieldArg(
+          oldType.ofType,
+          newType.ofType,
+        )
+      ) ||
+      // moving from non-null to nullable of the same underlying type is safe
+      (
+        !(newType instanceof GraphQLNonNull) &&
+        isChangeSafeForInputObjectFieldOrFieldArg(
+          oldType.ofType,
+          newType,
+        )
+      );
+  }
+  return isSafe;
+}
+
```

Comment:
I would take your suggestion if it was possible to share the `namedTypesMatch`/`listTypesMatch`/`nonNullTypesMatch` functions. But since `Interface/Object Types` and `Input Types` have slightly different notions of what's safe regarding nullability changes, it's not possible to share those functions and that means we'd just have 6 more helper functions.

I think it's better off as it is.